### PR TITLE
fixes bug: visible_name was not updated when not set in existing host…

### PR DESCRIPTION
**Pull request by raleonardo**
… in zabbix

##### SUMMARY
When updating visible_name for a host which has it unset, the update didn't  happen.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
zabbix_host

##### ADDITIONAL INFORMATION
The second condition was changed to validate that the new visible_name is not the same as the technical host name already set in zabbix
```

### CODE REVIEW

File: changelogs/fragments/fix_zabbix_host_visible_name.yml
Comment (D3DeFi) [inline changelogs/fragments/fix_zabbix_host_visible_name.yml]:
```suggestion
    - zabbix_host - was not possible to update a host where visible_name was not set in zabbix
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_host.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_host.py]: Please let me know what do you think about this:

```suggestion
            if host['name'] != visible_name:
```

**Review by D3DeFi — CHANGES_REQUESTED** [review]
Hi @raleonardo , no need to close PR and open a new one if CI is failing. It is usually enough to close & reopen PR that is failing, which then triggers CI rerun. 

Thank you for submitting this patch against devel branch. My review from #63854 still stands. At minimum, changelog fragment needs to specify a module which it targets.

File: lib/ansible/modules/monitoring/zabbix/zabbix_host.py
Comment (raleonardo) [reply lib/ansible/modules/monitoring/zabbix/zabbix_host.py]: for me it also works, but the second condition avoids  changing zabbix _visible name_ into the same as the _technical name_
But I guess this is not relevant if it is

File: changelogs/fragments/fix_zabbix_host_visible_name.yml
Comment (raleonardo) [reply changelogs/fragments/fix_zabbix_host_visible_name.yml]: thanks

**Review by D3DeFi — APPROVED** [review]
Thanks for the contribution @raleonardo 
!needs_revision
shipit

**Review by D3DeFi — APPROVED** [review]
shipit

**Comment by raleonardo** [comment]
@D3DeFi when would this be merged?

**Comment by D3DeFi** [comment]
> 
> 
> @D3DeFi when would this be merged?

not sure, every PR needs 2 shipit votes from different module or core maintainers

cc @rubentsirunyan

**Comment by D3DeFi** [comment]
cc @sky-joker

**Comment by sky-joker** [comment]
Thank you @raleonardo for this patch :)

I confirmed that `visible_name` can be set using this patch applied module, but I could not delete it.
Create a host and after running the following playbook the `visible_name` setting has been changed.

```yaml
---
- name: zabbix_host module test
  hosts: localhost
  gather_facts: no
  vars:
    zbx_url: zabbix server url
    zbx_user: admin
    zbx_password: password
  tasks:
    - name: test task for zabbix_host module
      zabbix_host:
        server_url: "{{ zbx_url }}"
        login_user: "{{ zbx_user }}"
        login_password: "{{ zbx_password }}"
        host_name: TEST
        visible_name: TESTName
        host_groups:
          - Linux servers
        interfaces:
          - type: 1
            main: 1
            useip: 1
            ip: 10.1.0.1
            dns: ""
            port: 10050
```

I tried to remove the `visible_name` setting and ran the following playbook but the setting was not removed.

```yaml
---
- name: zabbix_host module test
  hosts: localhost
  gather_facts: no
  vars:
    zbx_url: zabbix server url
    zbx_user: admin
    zbx_password: password
  tasks:
    - name: test task for zabbix_host module
      zabbix_host:
        server_url: "{{ zbx_url }}"
        login_user: "{{ zbx_user }}"
        login_password: "{{ zbx_password }}"
        host_name: TEST
        visible_name: ''
        host_groups:
          - Linux servers
        interfaces:
          - type: 1
            main: 1
            useip: 1
            ip: 10.1.0.1
            dns: ""
            port: 10050
```

@D3DeFi

What do you think about removing the `visible_name` setting?

**Comment by sky-joker** [comment]
Thank you @raleonardo for this patch :)

I confirmed that `visible_name` can be set using this patch applied module, but I could not delete it.
Create a host and after running the following playbook the `visible_name` setting has been changed.

```yaml
---
- name: zabbix_host module test
  hosts: localhost
  gather_facts: no
  vars:
    zbx_url: zabbix server url
    zbx_user: admin
    zbx_password: password
  tasks:
    - name: test task for zabbix_host module
      zabbix_host:
        server_url: "{{ zbx_url }}"
        login_user: "{{ zbx_user }}"
        login_password: "{{ zbx_password }}"
        host_name: TEST
        visible_name: TESTName
        host_groups:
          - Linux servers
        interfaces:
          - type: 1
            main: 1
            useip: 1
            ip: 10.1.0.1
            dns: ""
            port: 10050
```

I tried to remove the `visible_name` setting and ran the following playbook but the setting was not removed.

```yaml
---
- name: zabbix_host module test
  hosts: localhost
  gather_facts: no
  vars:
    zbx_url: zabbix server url
    zbx_user: admin
    zbx_password: password
  tasks:
    - name: test task for zabbix_host module
      zabbix_host:
        server_url: "{{ zbx_url }}"
        login_user: "{{ zbx_user }}"
        login_password: "{{ zbx_password }}"
        host_name: TEST
        visible_name: ''
        host_groups:
          - Linux servers
        interfaces:
          - type: 1
            main: 1
            useip: 1
            ip: 10.1.0.1
            dns: ""
            port: 10050
```

@D3DeFi

What do you think about removing the `visible_name` setting?

**Comment by D3DeFi** [comment]
Not sure if I remember this correctly, but the last time I've reviewed this PR, I've tried to play with raw python zabbix-api module to see if something like `zbx.host.update({....., 'visible_name': '', .... })` is possible and Zabbix server probably just ignored that parameter altogether. Need to test it again when I have the chance

**Comment by sky-joker** [comment]
The following modifications were required to remove `visible_name`.

```patch
--- before/zabbix_host.py   2019-12-14 19:01:27.500791213 +0900
+++ after/zabbix_host.py    2019-12-14 19:01:18.329608282 +0900
@@ -364,7 +364,7 @@
         try:
             if self._module.check_mode:
                 self._module.exit_json(changed=True)
-            parameters = {'hostid': host_id, 'groups': group_ids, 'status': status, 'tls_connect': tls_connect,
+            parameters = {'hostid': host_id, 'host': host_name, 'groups': group_ids, 'status': status, 'tls_connect': tls_connect,
                           'tls_accept': tls_accept}
             if proxy_id >= 0:
                 parameters['proxy_hostid'] = proxy_id
@@ -532,8 +532,10 @@
             return True

         # Check whether the visible_name has changed; Zabbix defaults to the technical hostname if not set.
-        if visible_name:
-            if host['name'] != visible_name:
+        if visible_name or visible_name == '':
+            if visible_name == '' and host['name'] != host['host']:
+                return True
+            elif host['name'] != visible_name and visible_name != '':
                 return True

         # Only compare description if it is given as a module parameter
```

I used Zabbix 4.0 for testing.
@D3DeFi, Could you check it?

**Comment by D3DeFi** [comment]
> 
> 
> The following modifications were required to remove `visible_name`.
> 
> ```diff
> --- before/zabbix_host.py   2019-12-14 19:01:27.500791213 +0900
> +++ after/zabbix_host.py    2019-12-14 19:01:18.329608282 +0900
> @@ -364,7 +364,7 @@
>          try:
>              if self._module.check_mode:
>                  self._module.exit_json(changed=True)
> -            parameters = {'hostid': host_id, 'groups': group_ids, 'status': status, 'tls_connect': tls_connect,
> +            parameters = {'hostid': host_id, 'host': host_name, 'groups': group_ids, 'status': status, 'tls_connect': tls_connect,
>                            'tls_accept': tls_accept}
>              if proxy_id >= 0:
>                  parameters['proxy_hostid'] = proxy_id
> @@ -532,8 +532,10 @@
>              return True
> 
>          # Check whether the visible_name has changed; Zabbix defaults to the technical hostname if not set.
> -        if visible_name:
> -            if host['name'] != visible_name:
> +        if visible_name or visible_name == '':
> +            if visible_name == '' and host['name'] != host['host']:
> +                return True
> +            elif host['name'] != visible_name and visible_name != '':
>                  return True
> 
>          # Only compare description if it is given as a module parameter
> ```
> 
> I used Zabbix 4.0 for testing.
> @D3DeFi, Could you check it?

I think it would miss change from VisibleNameA to VisibleNameB or atleast report it as not changed even though `host.update` will change `visible_name` to VisibleNameB

**Comment by sky-joker** [comment]
@D3DeFi , Thank you for the check.
I can't find the solution right away, so would you like to put this matter on hold and merge this patch?

**Comment by D3DeFi** [comment]
Yup, lets merge this. It is better to have a solution that will fix 90% of the cases than have it mostly bugged. Setting visible_name to empty string is probably a very rare case. Not saying that Zabbix will set visible_name to host_name in the background, so user can achieve the same result by simply setting visible_name to host_name in playbook :)

**Comment by sky-joker** [comment]
okay :)
shipit

**Comment by D3DeFi** [comment]
Thank you all! :)

**Pull request merged**
